### ISSUE TITLE
Symbols, Object.assign etc.

### ISSUE THREAD

**Issue by origin1tech** [TypeStrong/ts-node#383]
Compiling via Typescript compiler has no issues, however in tests for example Object.assign is not available whereas Typescript will compiling it just fine.

Disclosure haven't used ts-node all that much usually compile then run tests in workflow but curious what I'm missing here.

The error I'm getting for Object.assign here below is obvious as to the what it is, the why it can't compile yet Typescript can isn't obvious to me.

_Property 'assign' does not exist on type 'ObjectConstructor'._

**Comment by blakeembrey** [comment]
Can share your `tsconfig.json` file? This error specifically means you have an incorrect `lib` or `target` (probably using `es5` instead of `es2015`). I'm surprised it works using `tsc` though, since they should be using the same config.

**Comment by origin1tech** [comment]
@blakeembrey yeah I should have given more detail (I know better) was late last night..

<pre>
{
  "compilerOptions": {
    "declaration": false,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "lib": [
      "es6"
    ],
    "module": "commonjs",
    "moduleResolution": "node",
    "rootDir": "./",
    "outDir": "../dist",
    "sourceMap": false,
    "target": "es6",
    "typeRoots": [
      "../node_modules/@types"
    ],
    "types": [
      "node"
    ]
  },
  "include": [
    "./**/*.spec.ts"
  ]
}
</pre>

I'm referencing this by calling mocha via npm scripts which points to a mocha.opts file which then references the following script to hand off options, project config etc to ts-node.

<pre>
/**
 * TS-Node
 * Helper file to setup ts-node for testing.
 * @see https://github.com/TypeStrong/ts-node#configuration-options
 */

const tsnode = require('ts-node');

// Options
const opts = {
  project: './src/tsconfig.spec.json',
  ignoreWarnings: true,
  disableWarnings: true
};

// Register ts-node.
tsnode.register(opts);
</pre>

Lastly and while I have you if you have any suggestions glad to hear it. Like fastest options for testing. Not too concerned with warnings etc because I'll be compiling in the workflow for dist anyway and it will catch any errors directly from the Typescript compiler. So it seems to me at least in this scenario it makes sense to just get the quickest compile without too many warnings/nags, but I defer to you on that!!

**Comment by blakeembrey** [comment]
That seems odd. It looks right. If you wanted to create a repro, I'd gladly take a look. However, it seems like correcting the errors don't really matter - in which case you can just use "fast" mode and it'll only be a transpiler - no type checking.

**Comment by origin1tech** [comment]
Yeah  I saw the "fast" option. 

I thought the workflow was right but for some reason it still bombs on Object.assign, baffles me because if I compile via Typescript directly even to es5 as target zero issues. 

Incidentally I get argument expect errors too like expected 1 arg but got two which again is odd as that doesn't seem accurate and doesn't bomb when compiling via TS directly.

So I guess table this and I'll push up the repo here (its just a helper lib) later today or tomorrow, just finishing up the test.

Iike to get the kinks worked out as this is a nice little workflow as it prevents need for a bunch of babel transpiler packages, just your types, Typescript & ts-node.

**Comment by origin1tech** [comment]
@blakeembrey ahhhhh it's #istanbul that's causing the issue! Just dumb thumbed it and didn't take note of "nyc's" default require statement. I need to be calling my actual custom script and not **ts-node/register** directly.

NOTE: Blake if you want me to do a PR and add this to a section in the readme or you just want to stick it elsewhere happy to take care of that...

So for posterity sake if you're calling **mocha** via npm scripts as follows.

<pre>
{
  "test": "nyc mocha --opts ./src/mocha.opts"
}
</pre>

Your **nyc (Istanbul) config in package.json** would need to be as follows:

<pre>
  "nyc": {
    "include": [
      "src/**/*.ts",
      "src/**/*.tsx"
    ],
    "exclude": [
      "src/**/*.spec.ts"
    ],
    "extension": [
      ".ts",
      ".tsx"
    ],
    "require": [
      "./build/spec"
    ],
    "reporter": [
      "text",
      "html"
    ],
    "sourceMap": true,
    "instrument": true
  }
</pre>

Take note of the **"require"** property. In order to have mocha use a specific config (in my case tsconfig.spec.json) you need to create a script to fire up ts-node along with the project's (the tsconfig) path and any ts-node specific options. 

Both mocha and the above require in your nyc (Istanbul) config appear to be required.

**Comment by blakeembrey** [comment]
Interesting. If there's a succinct way to put it in the README, go for it. It's odd since I believe a few other people have been using `nyc` without an issue. Might be good to create an `examples/` folder with these workings as integration tests.

**Comment by origin1tech** [comment]
Right...well maybe give it a thought and postback once you've decided how you want to address it. 

Side note what's interesting is or should say what tripped me up was I thought well let me rename the tsconfig, make sure it bombs using the right one yada yada. Sure enough it bombs so I was thinking well it's for sure getting the right config and "es6" was set in lib so it should compile just as it would if I call tsc directly. But no joy.

Think the reason it works for others as you mentioned is tsconfig.json is likely in default location in project, mocha.opts is probably in test/mocha.opts.

I don't care for that. We like to keep tests in same as our source and just exclude them from our main tsconfig then you can have say /home/home.spec.ts /about/about.spec.ts and so on. far easier to find etc. 

So basically this workflow is nyc config where you require your custom script that sets the project for ts-node and you **don't** need the following in mocha opts:

<pre>
--compilers ts:./build/spec,tsx:./build/spec
</pre>

If not using nyc you would of course need that flag in your mocha.opts.

**Anyway just holler when you decide how you want to go about this. More than willing help out.**

**Comment by origin1tech** [comment]
...and I'll be done with this lib tomorrow so I'll push it and then you can just pull and see exactly the setup as well as play around if you wish.

**Comment by origin1tech** [comment]
@blakeembrey you want to close this? All is working well here, let me know if you want me to PR any docs on this. You can see how this is working at https://github.com/origin1tech/chek 

Key things to look at here are:

"nyc" property in package.json
"spec.js" in build/spec.js
"tsconfig.json" & "tsconfig.spec.json" in src/
"mocha.opts" in src/
"test" property in package.json "scripts" key.

Hope this saves someone else some time :)

**Comment by blakeembrey** [comment]
Thanks. Feel free to create a PR with an `examples/nyc` folder. Maybe we can add it to the test suite in the future.
